File: storage.py
"""
Local Storage module for Idea Vault - Browser-based persistence using streamlit-local-storage.
Data is stored in the browser's localStorage, private to each device/browser.
"""
import streamlit as st
from streamlit_local_storage import LocalStorage
from datetime import datetime
from typing import Optional
import json

# Initialize local storage with a unique key prefix
_local_storage = None

def _get_storage() -> LocalStorage:
    """Get or create the LocalStorage instance."""
    global _local_storage
    if _local_storage is None:
        _local_storage = LocalStorage()
    return _local_storage


def _get_data(key: str, default=None):
    """Get data from local storage."""
    storage = _get_storage()
    try:
        value = storage.getItem(key)
        if value is None:
            return default
        return json.loads(value) if isinstance(value, str) else value
    except Exception:
        return default


def _set_data(key: str, value):
    """Set data in local storage."""
    storage = _get_storage()
    try:
        storage.setItem(key, json.dumps(value))
    except Exception as e:
        st.error(f"Failed to save data: {e}")


def _get_next_id(key: str) -> int:
    """Get the next available ID for a collection."""
    counter_key = f"ideavault_counter_{key}"
    current = _get_data(counter_key, 0)
    next_id = current + 1
    _set_data(counter_key, next_id)
    return next_id


# =============================================================================
# INITIALIZATION
# =============================================================================

def init_db():
    """Initialize storage structure if needed (no-op for localStorage, but ensures keys exist)."""
    # Initialize empty arrays if they don't exist
    if _get_data("ideavault_problems") is None:
        _set_data("ideavault_problems", [])
    if _get_data("ideavault_ideas") is None:
        _set_data("ideavault_ideas", [])
    if _get_data("ideavault_notes") is None:
        _set_data("ideavault_notes", [])
    if _get_data("ideavault_problem_idea_links") is None:
        _set_data("ideavault_problem_idea_links", [])


# =============================================================================
# PROBLEMS CRUD
# =============================================================================

def create_problem(title: str, description: str = "", observed_context: str = "",
                   severity: int = 3, frequency: str = "weekly", status: str = "open",
                   tags: str = "") -> int:
    """Create a new problem and return its ID."""
    problems = _get_data("ideavault_problems", [])
    now = datetime.now().isoformat()
    problem_id = _get_next_id("problems")

    problem = {
        "id": problem_id,
        "title": title,
        "description": description,
        "observed_context": observed_context,
        "severity": severity,
        "frequency": frequency,
        "status": status,
        "tags": tags,
        "created_at": now,
        "updated_at": now
    }

    problems.append(problem)
    _set_data("ideavault_problems", problems)
    return problem_id


def get_all_problems() -> list[dict]:
    """Get all problems as a list of dicts."""
    problems = _get_data("ideavault_problems", [])
    # Sort by created_at descending
    return sorted(problems, key=lambda x: x.get('created_at', ''), reverse=True)


def get_problem_by_id(problem_id: int) -> Optional[dict]:
    """Get a single problem by ID."""
    problems = _get_data("ideavault_problems", [])
    for p in problems:
        if p['id'] == problem_id:
            return p
    return None


def update_problem(problem_id: int, title: str, description: str, observed_context: str,
                   severity: int, frequency: str, status: str, tags: str) -> bool:
    """Update an existing problem."""
    problems = _get_data("ideavault_problems", [])
    now = datetime.now().isoformat()

    for i, p in enumerate(problems):
        if p['id'] == problem_id:
            problems[i].update({
                "title": title,
                "description": description,
                "observed_context": observed_context,
                "severity": severity,
                "frequency": frequency,
                "status": status,
                "tags": tags,
                "updated_at": now
            })
            _set_data("ideavault_problems", problems)
            return True
    return False


def delete_problem(problem_id: int) -> bool:
    """Delete a problem by ID (also removes related links and nullifies note references)."""
    problems = _get_data("ideavault_problems", [])
    original_len = len(problems)
    problems = [p for p in problems if p['id'] != problem_id]

    if len(problems) < original_len:
        _set_data("ideavault_problems", problems)

        # Remove related links (cascade)
        links = _get_data("ideavault_problem_idea_links", [])
        links = [l for l in links if l['problem_id'] != problem_id]
        _set_data("ideavault_problem_idea_links", links)

        # Nullify note references (SET NULL behavior)
        notes = _get_data("ideavault_notes", [])
        for note in notes:
            if note.get('problem_id') == problem_id:
                note['problem_id'] = None
        _set_data("ideavault_notes", notes)

        return True
    return False


def get_problems_count_by_status() -> dict:
    """Get count of problems grouped by status."""
    problems = _get_data("ideavault_problems", [])
    counts = {}
    for p in problems:
        status = p.get('status', 'open')
        counts[status] = counts.get(status, 0) + 1
    return counts


def get_recent_problems(limit: int = 5) -> list[dict]:
    """Get most recently added problems."""
    problems = get_all_problems()
    return problems[:limit]


# =============================================================================
# IDEAS CRUD
# =============================================================================

def create_idea(title: str, pitch: str = "", target_user: str = "", value_prop: str = "",
                differentiation: str = "", assumptions: str = "", risks: str = "",
                status: str = "new", score: Optional[int] = None, tags: str = "") -> int:
    """Create a new idea and return its ID."""
    ideas = _get_data("ideavault_ideas", [])
    now = datetime.now().isoformat()
    idea_id = _get_next_id("ideas")

    idea = {
        "id": idea_id,
        "title": title,
        "pitch": pitch,
        "target_user": target_user,
        "value_prop": value_prop,
        "differentiation": differentiation,
        "assumptions": assumptions,
        "risks": risks,
        "status": status,
        "score": score,
        "tags": tags,
        "created_at": now,
        "updated_at": now
    }

    ideas.append(idea)
    _set_data("ideavault_ideas", ideas)
    return idea_id


def get_all_ideas() -> list[dict]:
    """Get all ideas as a list of dicts."""
    ideas = _get_data("ideavault_ideas", [])
    return sorted(ideas, key=lambda x: x.get('created_at', ''), reverse=True)


def get_idea_by_id(idea_id: int) -> Optional[dict]:
    """Get a single idea by ID."""
    ideas = _get_data("ideavault_ideas", [])
    for i in ideas:
        if i['id'] == idea_id:
            return i
    return None


def update_idea(idea_id: int, title: str, pitch: str, target_user: str, value_prop: str,
                differentiation: str, assumptions: str, risks: str, status: str,
                score: Optional[int], tags: str) -> bool:
    """Update an existing idea."""
    ideas = _get_data("ideavault_ideas", [])
    now = datetime.now().isoformat()

    for i, idea in enumerate(ideas):
        if idea['id'] == idea_id:
            ideas[i].update({
                "title": title,
                "pitch": pitch,
                "target_user": target_user,
                "value_prop": value_prop,
                "differentiation": differentiation,
                "assumptions": assumptions,
                "risks": risks,
                "status": status,
                "score": score,
                "tags": tags,
                "updated_at": now
            })
            _set_data("ideavault_ideas", ideas)
            return True
    return False


def delete_idea(idea_id: int) -> bool:
    """Delete an idea by ID (also removes related links and nullifies note references)."""
    ideas = _get_data("ideavault_ideas", [])
    original_len = len(ideas)
    ideas = [i for i in ideas if i['id'] != idea_id]

    if len(ideas) < original_len:
        _set_data("ideavault_ideas", ideas)

        # Remove related links (cascade)
        links = _get_data("ideavault_problem_idea_links", [])
        links = [l for l in links if l['idea_id'] != idea_id]
        _set_data("ideavault_problem_idea_links", links)

        # Nullify note references (SET NULL behavior)
        notes = _get_data("ideavault_notes", [])
        for note in notes:
            if note.get('idea_id') == idea_id:
                note['idea_id'] = None
        _set_data("ideavault_notes", notes)

        return True
    return False


def get_ideas_count_by_status() -> dict:
    """Get count of ideas grouped by status."""
    ideas = _get_data("ideavault_ideas", [])
    counts = {}
    for i in ideas:
        status = i.get('status', 'new')
        counts[status] = counts.get(status, 0) + 1
    return counts


def get_recent_ideas(limit: int = 5) -> list[dict]:
    """Get most recently added ideas."""
    ideas = get_all_ideas()
    return ideas[:limit]


# =============================================================================
# PROBLEM-IDEA LINKS
# =============================================================================

def link_problem_to_idea(problem_id: int, idea_id: int) -> bool:
    """Create a link between a problem and an idea."""
    links = _get_data("ideavault_problem_idea_links", [])

    # Check if link already exists
    for l in links:
        if l['problem_id'] == problem_id and l['idea_id'] == idea_id:
            return False  # Already exists

    link_id = _get_next_id("links")
    links.append({
        "id": link_id,
        "problem_id": problem_id,
        "idea_id": idea_id
    })
    _set_data("ideavault_problem_idea_links", links)
    return True


def unlink_problem_from_idea(problem_id: int, idea_id: int) -> bool:
    """Remove a link between a problem and an idea."""
    links = _get_data("ideavault_problem_idea_links", [])
    original_len = len(links)
    links = [l for l in links if not (l['problem_id'] == problem_id and l['idea_id'] == idea_id)]

    if len(links) < original_len:
        _set_data("ideavault_problem_idea_links", links)
        return True
    return False


def get_ideas_for_problem(problem_id: int) -> list[dict]:
    """Get all ideas linked to a problem."""
    links = _get_data("ideavault_problem_idea_links", [])
    idea_ids = [l['idea_id'] for l in links if l['problem_id'] == problem_id]

    ideas = _get_data("ideavault_ideas", [])
    return [i for i in ideas if i['id'] in idea_ids]


def get_problems_for_idea(idea_id: int) -> list[dict]:
    """Get all problems linked to an idea."""
    links = _get_data("ideavault_problem_idea_links", [])
    problem_ids = [l['problem_id'] for l in links if l['idea_id'] == idea_id]

    problems = _get_data("ideavault_problems", [])
    return [p for p in problems if p['id'] in problem_ids]


def get_linked_problem_ids_for_idea(idea_id: int) -> list[int]:
    """Get IDs of all problems linked to an idea."""
    links = _get_data("ideavault_problem_idea_links", [])
    return [l['problem_id'] for l in links if l['idea_id'] == idea_id]


def set_problem_links_for_idea(idea_id: int, problem_ids: list[int]):
    """Set the problem links for an idea (replaces existing links)."""
    links = _get_data("ideavault_problem_idea_links", [])

    # Remove existing links for this idea
    links = [l for l in links if l['idea_id'] != idea_id]

    # Add new links
    for pid in problem_ids:
        link_id = _get_next_id("links")
        links.append({
            "id": link_id,
            "problem_id": pid,
            "idea_id": idea_id
        })

    _set_data("ideavault_problem_idea_links", links)


# =============================================================================
# NOTES CRUD
# =============================================================================

def create_note(note_type: str, content: str, links: str = "",
                problem_id: Optional[int] = None, idea_id: Optional[int] = None) -> int:
    """Create a new note and return its ID."""
    notes = _get_data("ideavault_notes", [])
    now = datetime.now().isoformat()
    note_id = _get_next_id("notes")

    note = {
        "id": note_id,
        "note_type": note_type,
        "content": content,
        "links": links,
        "problem_id": problem_id,
        "idea_id": idea_id,
        "created_at": now
    }

    notes.append(note)
    _set_data("ideavault_notes", notes)
    return note_id


def get_all_notes() -> list[dict]:
    """Get all notes as a list of dicts."""
    notes = _get_data("ideavault_notes", [])
    return sorted(notes, key=lambda x: x.get('created_at', ''), reverse=True)


def get_note_by_id(note_id: int) -> Optional[dict]:
    """Get a single note by ID."""
    notes = _get_data("ideavault_notes", [])
    for n in notes:
        if n['id'] == note_id:
            return n
    return None


def update_note(note_id: int, note_type: str, content: str, links: str,
                problem_id: Optional[int], idea_id: Optional[int]) -> bool:
    """Update an existing note."""
    notes = _get_data("ideavault_notes", [])

    for i, n in enumerate(notes):
        if n['id'] == note_id:
            notes[i].update({
                "note_type": note_type,
                "content": content,
                "links": links,
                "problem_id": problem_id,
                "idea_id": idea_id
            })
            _set_data("ideavault_notes", notes)
            return True
    return False


def delete_note(note_id: int) -> bool:
    """Delete a note by ID."""
    notes = _get_data("ideavault_notes", [])
    original_len = len(notes)
    notes = [n for n in notes if n['id'] != note_id]

    if len(notes) < original_len:
        _set_data("ideavault_notes", notes)
        return True
    return False


def get_notes_for_problem(problem_id: int) -> list[dict]:
    """Get all notes attached to a problem."""
    notes = _get_data("ideavault_notes", [])
    result = [n for n in notes if n.get('problem_id') == problem_id]
    return sorted(result, key=lambda x: x.get('created_at', ''), reverse=True)


def get_notes_for_idea(idea_id: int) -> list[dict]:
    """Get all notes attached to an idea."""
    notes = _get_data("ideavault_notes", [])
    result = [n for n in notes if n.get('idea_id') == idea_id]
    return sorted(result, key=lambda x: x.get('created_at', ''), reverse=True)


# =============================================================================
# DATA EXPORT/IMPORT (for backup/restore)
# =============================================================================

def export_all_data() -> dict:
    """Export all data for backup purposes."""
    return {
        "problems": _get_data("ideavault_problems", []),
        "ideas": _get_data("ideavault_ideas", []),
        "notes": _get_data("ideavault_notes", []),
        "links": _get_data("ideavault_problem_idea_links", []),
        "counters": {
            "problems": _get_data("ideavault_counter_problems", 0),
            "ideas": _get_data("ideavault_counter_ideas", 0),
            "notes": _get_data("ideavault_counter_notes", 0),
            "links": _get_data("ideavault_counter_links", 0),
        }
    }


def import_all_data(data: dict) -> bool:
    """Import data from a backup."""
    try:
        if "problems" in data:
            _set_data("ideavault_problems", data["problems"])
        if "ideas" in data:
            _set_data("ideavault_ideas", data["ideas"])
        if "notes" in data:
            _set_data("ideavault_notes", data["notes"])
        if "links" in data:
            _set_data("ideavault_problem_idea_links", data["links"])
        if "counters" in data:
            for key, val in data["counters"].items():
                _set_data(f"ideavault_counter_{key}", val)
        return True
    except Exception:
        return False


def clear_all_data():
    """Clear all data from local storage."""
    _set_data("ideavault_problems", [])
    _set_data("ideavault_ideas", [])
    _set_data("ideavault_notes", [])
    _set_data("ideavault_problem_idea_links", [])
    _set_data("ideavault_counter_problems", 0)
    _set_data("ideavault_counter_ideas", 0)
    _set_data("ideavault_counter_notes", 0)
    _set_data("ideavault_counter_links", 0)

File: src/lib/storage.ts
import { Problem, Idea, Note, ProblemIdeaLink, StorageData } from './types';

const STORAGE_KEY = 'ideavault_data';

function getStorageData(): StorageData {
  if (typeof window === 'undefined') {
    return {
      problems: [],
      ideas: [],
      notes: [],
      links: [],
      counters: { problems: 0, ideas: 0, notes: 0, links: 0 }
    };
  }

  const stored = localStorage.getItem(STORAGE_KEY);
  if (!stored) {
    const initial: StorageData = {
      problems: [],
      ideas: [],
      notes: [],
      links: [],
      counters: { problems: 0, ideas: 0, notes: 0, links: 0 }
    };
    localStorage.setItem(STORAGE_KEY, JSON.stringify(initial));
    return initial;
  }
  return JSON.parse(stored);
}

function saveStorageData(data: StorageData): void {
  if (typeof window === 'undefined') return;
  localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
}

function getNextId(data: StorageData, type: keyof StorageData['counters']): number {
  data.counters[type]++;
  return data.counters[type];
}

// =============================================================================
// PROBLEMS CRUD
// =============================================================================

export function createProblem(
  title: string,
  description = '',
  observed_context = '',
  severity = 3,
  frequency: Problem['frequency'] = 'weekly',
  status: Problem['status'] = 'open',
  tags = ''
): number {
  const data = getStorageData();
  const now = new Date().toISOString();
  const id = getNextId(data, 'problems');

  const problem: Problem = {
    id,
    title,
    description,
    observed_context,
    severity,
    frequency,
    status,
    tags,
    created_at: now,
    updated_at: now
  };

  data.problems.push(problem);
  saveStorageData(data);
  return id;
}

export function getAllProblems(): Problem[] {
  const data = getStorageData();
  return [...data.problems].sort((a, b) =>
    new Date(b.created_at).getTime() - new Date(a.created_at).getTime()
  );
}

export function getProblemById(id: number): Problem | null {
  const data = getStorageData();
  return data.problems.find(p => p.id === id) || null;
}

export function updateProblem(
  id: number,
  title: string,
  description: string,
  observed_context: string,
  severity: number,
  frequency: Problem['frequency'],
  status: Problem['status'],
  tags: string
): boolean {
  const data = getStorageData();
  const index = data.problems.findIndex(p => p.id === id);

  if (index === -1) return false;

  data.problems[index] = {
    ...data.problems[index],
    title,
    description,
    observed_context,
    severity,
    frequency,
    status,
    tags,
    updated_at: new Date().toISOString()
  };

  saveStorageData(data);
  return true;
}

export function deleteProblem(id: number): boolean {
  const data = getStorageData();
  const originalLen = data.problems.length;
  data.problems = data.problems.filter(p => p.id !== id);

  if (data.problems.length < originalLen) {
    // Cascade delete links
    data.links = data.links.filter(l => l.problem_id !== id);
    // Nullify note references
    data.notes = data.notes.map(n =>
      n.problem_id === id ? { ...n, problem_id: null } : n
    );
    saveStorageData(data);
    return true;
  }
  return false;
}

export function getProblemsCountByStatus(): Record<string, number> {
  const data = getStorageData();
  const counts: Record<string, number> = {};
  data.problems.forEach(p => {
    counts[p.status] = (counts[p.status] || 0) + 1;
  });
  return counts;
}

export function getRecentProblems(limit = 5): Problem[] {
  return getAllProblems().slice(0, limit);
}

// =============================================================================
// IDEAS CRUD
// =============================================================================

export function createIdea(
  title: string,
  pitch = '',
  target_user = '',
  value_prop = '',
  differentiation = '',
  assumptions = '',
  risks = '',
  status: Idea['status'] = 'new',
  score: number | null = null,
  tags = ''
): number {
  const data = getStorageData();
  const now = new Date().toISOString();
  const id = getNextId(data, 'ideas');

  const idea: Idea = {
    id,
    title,
    pitch,
    target_user,
    value_prop,
    differentiation,
    assumptions,
    risks,
    status,
    score,
    tags,
    created_at: now,
    updated_at: now
  };

  data.ideas.push(idea);
  saveStorageData(data);
  return id;
}

export function getAllIdeas(): Idea[] {
  const data = getStorageData();
  return [...data.ideas].sort((a, b) =>
    new Date(b.created_at).getTime() - new Date(a.created_at).getTime()
  );
}

export function getIdeaById(id: number): Idea | null {
  const data = getStorageData();
  return data.ideas.find(i => i.id === id) || null;
}

export function updateIdea(
  id: number,
  title: string,
  pitch: string,
  target_user: string,
  value_prop: string,
  differentiation: string,
  assumptions: string,
  risks: string,
  status: Idea['status'],
  score: number | null,
  tags: string
): boolean {
  const data = getStorageData();
  const index = data.ideas.findIndex(i => i.id === id);

  if (index === -1) return false;

  data.ideas[index] = {
    ...data.ideas[index],
    title,
    pitch,
    target_user,
    value_prop,
    differentiation,
    assumptions,
    risks,
    status,
    score,
    tags,
    updated_at: new Date().toISOString()
  };

  saveStorageData(data);
  return true;
}

export function deleteIdea(id: number): boolean {
  const data = getStorageData();
  const originalLen = data.ideas.length;
  data.ideas = data.ideas.filter(i => i.id !== id);

  if (data.ideas.length < originalLen) {
    // Cascade delete links
    data.links = data.links.filter(l => l.idea_id !== id);
    // Nullify note references
    data.notes = data.notes.map(n =>
      n.idea_id === id ? { ...n, idea_id: null } : n
    );
    saveStorageData(data);
    return true;
  }
  return false;
}

export function getIdeasCountByStatus(): Record<string, number> {
  const data = getStorageData();
  const counts: Record<string, number> = {};
  data.ideas.forEach(i => {
    counts[i.status] = (counts[i.status] || 0) + 1;
  });
  return counts;
}

export function getRecentIdeas(limit = 5): Idea[] {
  return getAllIdeas().slice(0, limit);
}

// =============================================================================
// PROBLEM-IDEA LINKS
// =============================================================================

export function linkProblemToIdea(problem_id: number, idea_id: number): boolean {
  const data = getStorageData();

  // Check if link already exists
  if (data.links.some(l => l.problem_id === problem_id && l.idea_id === idea_id)) {
    return false;
  }

  const id = getNextId(data, 'links');
  data.links.push({ id, problem_id, idea_id });
  saveStorageData(data);
  return true;
}

export function unlinkProblemFromIdea(problem_id: number, idea_id: number): boolean {
  const data = getStorageData();
  const originalLen = data.links.length;
  data.links = data.links.filter(l =>
    !(l.problem_id === problem_id && l.idea_id === idea_id)
  );

  if (data.links.length < originalLen) {
    saveStorageData(data);
    return true;
  }
  return false;
}

export function getIdeasForProblem(problem_id: number): Idea[] {
  const data = getStorageData();
  const ideaIds = data.links
    .filter(l => l.problem_id === problem_id)
    .map(l => l.idea_id);
  return data.ideas.filter(i => ideaIds.includes(i.id));
}

export function getProblemsForIdea(idea_id: number): Problem[] {
  const data = getStorageData();
  const problemIds = data.links
    .filter(l => l.idea_id === idea_id)
    .map(l => l.problem_id);
  return data.problems.filter(p => problemIds.includes(p.id));
}

export function getLinkedProblemIdsForIdea(idea_id: number): number[] {
  const data = getStorageData();
  return data.links
    .filter(l => l.idea_id === idea_id)
    .map(l => l.problem_id);
}

export function setProblemLinksForIdea(idea_id: number, problem_ids: number[]): void {
  const data = getStorageData();

  // Remove existing links for this idea
  data.links = data.links.filter(l => l.idea_id !== idea_id);

  // Add new links
  problem_ids.forEach(problem_id => {
    const id = getNextId(data, 'links');
    data.links.push({ id, problem_id, idea_id });
  });

  saveStorageData(data);
}

// =============================================================================
// NOTES CRUD
// =============================================================================

export function createNote(
  note_type: Note['note_type'],
  content: string,
  links = '',
  problem_id: number | null = null,
  idea_id: number | null = null
): number {
  const data = getStorageData();
  const now = new Date().toISOString();
  const id = getNextId(data, 'notes');

  const note: Note = {
    id,
    note_type,
    content,
    links,
    problem_id,
    idea_id,
    created_at: now
  };

  data.notes.push(note);
  saveStorageData(data);
  return id;
}

export function getAllNotes(): Note[] {
  const data = getStorageData();
  return [...data.notes].sort((a, b) =>
    new Date(b.created_at).getTime() - new Date(a.created_at).getTime()
  );
}

export function getNoteById(id: number): Note | null {
  const data = getStorageData();
  return data.notes.find(n => n.id === id) || null;
}

export function updateNote(
  id: number,
  note_type: Note['note_type'],
  content: string,
  links: string,
  problem_id: number | null,
  idea_id: number | null
): boolean {
  const data = getStorageData();
  const index = data.notes.findIndex(n => n.id === id);

  if (index === -1) return false;

  data.notes[index] = {
    ...data.notes[index],
    note_type,
    content,
    links,
    problem_id,
    idea_id
  };

  saveStorageData(data);
  return true;
}

export function deleteNote(id: number): boolean {
  const data = getStorageData();
  const originalLen = data.notes.length;
  data.notes = data.notes.filter(n => n.id !== id);

  if (data.notes.length < originalLen) {
    saveStorageData(data);
    return true;
  }
  return false;
}

export function getNotesForProblem(problem_id: number): Note[] {
  const data = getStorageData();
  return data.notes
    .filter(n => n.problem_id === problem_id)
    .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
}

export function getNotesForIdea(idea_id: number): Note[] {
  const data = getStorageData();
  return data.notes
    .filter(n => n.idea_id === idea_id)
    .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
}

// =============================================================================
// DATA MANAGEMENT
// =============================================================================

export function exportAllData(): StorageData {
  return getStorageData();
}

export function importAllData(data: StorageData): boolean {
  try {
    saveStorageData(data);
    return true;
  } catch {
    return false;
  }
}

export function clearAllData(): void {
  const initial: StorageData = {
    problems: [],
    ideas: [],
    notes: [],
    links: [],
    counters: { problems: 0, ideas: 0, notes: 0, links: 0 }
  };
  saveStorageData(initial);
}
